### DISCUSSION TITLE
Clarification between "pragma key = x'...'" and "pragma hexkey = ..."

### DISCUSSION THREAD

**Discussion by Willena** [utelle/SQLite3MultipleCiphers#83] [Q&A] [answered]
Hi ! @utelle 

Just a quick question: 
In the [1.3.10](https://github.com/utelle/SQLite3MultipleCiphers/releases/tag/v1.3.10) release, `pragma hexkey/hexrekey` was added.

Could you explain the differences between the "new" `pragma hexkey` and existing `pragma key` with raw values shown in the [documentation website, `pragma key` section](https://utelle.github.io/SQLite3MultipleCiphers/docs/configuration/config_sql_pragmas/#pragma-key) ? 

Raw key data examples (taken from the doc):
```
-- Example of a raw key for the SQLCipher scheme  (no salt)
PRAGMA key = "x'54686973206973206D792076657279207365637265742070617373776F72642E'";
-- Example of a raw key for the sqleet scheme (no salt)
PRAGMA key = 'raw:54686973206973206D792076657279207365637265742070617373776F72642E';
```

```
-- Example of a raw key for the SQLCipher scheme (with salt)
PRAGMA key = "x'54686973206973206D792076657279207365637265742070617373776F72642E2E73616C7479206B65792073616C742E'";
-- Example of a raw key for the sqleet scheme (with salt)
PRAGMA key = 'raw:54686973206973206D792076657279207365637265742070617373776F72642E2E73616C7479206B65792073616C742E';
```

**Comment by utelle** [comment, marked as answer]
> Just a quick question: In the [1.3.10](https://github.com/utelle/SQLite3MultipleCiphers/releases/tag/v1.3.10) release, `pragma hexkey/hexrekey` was added.

Yes, this new pragma was added to provide better compatibility with the [official SEE extension](https://www.sqlite.org/see/doc/release/www/readme.wiki) (see paragraph 6).  If a passphrase contains binary data, it is cumbersome to specify it as a string value, if at all possible. `PRAGMA hexkey=` allows to specify a hexadecimal encoded string that is intercally decoded to binary data (which are then used to derive the actual encryption key).

> Could you explain the differences between the "new" `pragma hexkey` and existing `pragma key` with raw values shown in the [documentation website, `pragma key` section](https://utelle.github.io/SQLite3MultipleCiphers/docs/configuration/config_sql_pragmas/#pragma-key) ?

The variants of `PRAGMA key=` with raw values were introduced to provide compatibility with _SQLCipher_ resp _sqleet_. Both allow to provide raw key data (without performing key derivation). Only the _SQLCipher_ and _sqleet_ cipher schemes support this option and both require a specific syntax for the key data (as described in the examples). That is, it is the user's responsibility to perform key derivation with methods of his own choice.

In contrast, `PRAGMA hexkey=` does **not** circumvent key derivation.

I have to admit that the new pragma should be documented.

**Reply by Willena** [reply]
Thanks for the clarification !
